src/include/btree_cmp_inline.h

```diff
@@ -166,6 +306,145 @@ __wt_compare_bounds(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_ITEM *key, u
     return (0);
 }
 
+#ifdef HAVE_X86INTRIN_H
+/*
+ * __lex_compare_skip_gt_16 --
+ *     Lexicographic comparison routine for data greater than 16 bytes, skipping leading bytes.
+ *     Returns: < 0 if user_item is lexicographically < tree_item = 0 if user_item is
+ *     lexicographically = tree_item > 0 if user_item is lexicographically > tree_item We use the
+ *     names "user" and "tree" so it's clear in the btree code which the application is looking at
+ *     when we call its comparison function.
+ */
+static WT_INLINE int
+__lex_compare_skip_gt_16(
+  const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp, size_t *matchp)
+{
+    __m128i res_eq, t, u;
+    uint64_t t64, u64;
+    int32_t eq_bits;
+    size_t match, final_match;
+    const uint8_t *tendp, *treep, *uendp, *userp;
+
+    match = *matchp;
+    uendp = ustartp + len;
+    tendp = tstartp + len;
+
+    /*
+     * Compare 16 bytes at a time until we find a difference or run out of 16 byte chunks to
+     * compare.
+     */
+    for (userp = ustartp + match, treep = tstartp + match; uendp - userp > WT_VECTOR_SIZE;
+         userp += WT_VECTOR_SIZE, treep += WT_VECTOR_SIZE) {
+        u = _mm_loadu_si128((const __m128i *)userp);
+        t = _mm_loadu_si128((const __m128i *)treep);
+        res_eq = _mm_cmpeq_epi8(u, t);
+        if ((eq_bits = _mm_movemask_epi8(res_eq)) != 65535) {
+            match = (size_t)(userp - ustartp);
+            goto final128;
+        }
+    }
+
+    /*
+     * Rewind until there is exactly 16 bytes left. We know we started with at least 16, so we are
+     * still in bound.
+     */
+    u = _mm_loadu_si128((const __m128i *)(uendp - WT_VECTOR_SIZE));
+    t = _mm_loadu_si128((const __m128i *)(tendp - WT_VECTOR_SIZE));
+    res_eq = _mm_cmpeq_epi8(u, t);
+    eq_bits = _mm_movemask_epi8(res_eq);
+
+final128:
+    if (eq_bits == 65535) {
+        *matchp = len;
+        return (lencmp);
+    } else {
+        final_match = (size_t)__builtin_ctz(~(uint32_t)eq_bits);
+        match += final_match;
+        *matchp = match;
+        if (final_match >= sizeof(uint64_t)) {
+            u64 = (uint64_t)_mm_extract_epi64(u, 1);
+            t64 = (uint64_t)_mm_extract_epi64(t, 1);
+        } else {
+            u64 = (uint64_t)_mm_extract_epi64(u, 0);
+            t64 = (uint64_t)_mm_extract_epi64(t, 0);
+        }
+
+#ifndef WORDS_BIGENDIAN
+        u64 = __wt_bswap64(u64);
+        t64 = __wt_bswap64(t64);
+#endif
+
+        return (u64 < t64 ? -1 : 1);
+    }
```

Comment:
This could just be `return ustartp[match] - tstartp[match];`, which is how glibc does it. I think the reason it makes sense is that it avoids an (moderately) unpredictable branch and loading from L1 cache (which both bytes should be in because we just read them) is a similar cost to moving from vector registers to GPRs on x86 (but not on arm).

Reply:
I'm not sure that this is correct. We are comparing things in lex order not signed integer order. If we use signed integer, the order is different. For example, if ustartp[match] is -1 and tstartp[match] is 1, we will return -2 but in the unsigned integer case, ustartp[match] should be larger than tstartp[match].

Reply:
Note that the start pointers are uint8_t. That means you will get 255 rather than -1. Due to the weird rules about integer promotion in C (you basically can't do any operations except load and store on types narrower than int), that will be zero-extended up to an int prior to doing the subtraction so it will return 254 which is >0. You can see this happening in the ASM.